### ISSUE TITLE
Enumerations

### ISSUE THREAD

**Issue by lassik** [pre-srfi/static-scheme#13]
R6RS has enumerations, and SRFI 209 (_Enumerations and Enum Sets_) is still open.

How are enumerations exported in Scheme? Each enum value is a symbol (or an identifier in R6RS?). Is each identifier exported? If so, does this mean we can have enums where some values are public and others private? That seems like a reasonable idea.

**Comment by lassik** [comment]
If we have `(define-union-type color  red green blue)` in Steme then we should have some way to map it to a Scheme enums.

**Comment by lassik** [comment]
One approach is to re-use R6RS `define-enumeration` in the static language. It can define the same types of value constructors as `define-union-type` does in the static language, but if such an enum is imported into dynamic Scheme somehow, it could translate into a R6RS/SRFI-209 enum type.

In any case, there will be situations where static and dynamic code want to share some of the same enums, so we should have some way to write them only in one place and avoid duplication.

**Comment by johnwcowan** [comment]
R6RS enums are symbols.  If you define them with `define-enumeration`, identifiers with the same names as the symbols are bound.

R7RS enums are objects of a single unique type, but each of them has a symbol associated with it.  A similar `define-enumeration` will be added to SRFI 209.

However, per other discussions `(define-union-type a b c)` refers to existing *types*.   Union types have nothing to do with enumerations, though you can add a specific enumeration type to a union type just like any other predicate-defined type.

**Comment by lassik** [comment]
> However, per other discussions `(define-union-type a b c)` refers to existing _types_. Union types have nothing to do with enumerations

Not the case in static FP languages:

* Haskell `data` and ML `datatype` define one new type and one or more new constructors for it.
* A union type is the standard textbook approach for making an enum type.

In dynamic Scheme we have so far thought `define-union-type` should take existing types (e.g. record types) for the value constructors, but coming from a static FP background, that is somewhat anomalous. Not sure what's best.

**Comment by lassik** [comment]
IMHO the confusions we have around these distinctions highlight that we should already start thinking about what the static-dynamic boundary protocol should be like.

**Comment by mnieper** [comment]
I agree that we should think deeply about algebraic data types first.

**Comment by johnwcowan** [comment]
Ah, I follow you now.  Yes, on the Steme side a datatype specifies one or more constructors with a variable number of arguments each.  On the Scheme side, each constructor is a record type, and the datatype as a whole is just a union of their predicates.  That's all straightforward.

But it would be nice if an existing Scheme record type could become a constructor.  Consider this:

```
(define-datatype Thing
  (thing1 String String) ; constructor with two arguments
  (thing2) ; nullary constructor
  thing3) ; record-type defined in Scheme
```

Now perhaps a Scheme enum-type should look to Steme like a datatype all of whose constructors are nullary, which was your point above.  But it seems to me pointless to implement this:

```
(define-datatype Color
  (red)
  (blue)
  (purple))
```
 as an enum-type on the Scheme side.

That still leaves open the question of how nullary constructors are to be implemented on the Scheme side.

**Comment by mnieper** [comment]
The mapping from union types on the Steme side and finite sets of pairwise disjoint predicates on the Scheme side is not that straight-forward. The reason is that union types on the Steme set should not overlap I guess, while finite sets of pairwise disjoint predicates can, of course, overlap.

**Comment by johnwcowan** [comment]
I'm glad you brought that up; I knew it was lurking there.  So I think this is something Steme has to enforce.  Currently it doesn't mean anything to import or export the identifier bound to a record type by define-record-type, because we have no idea (per SRFI 9) what it is.  If we are to be able to do this across library boundaries (and a fortiori between Scheme and Steme) we need to have a better understanding of it,  I think.

**Comment by mnieper** [comment]
For R6RS (syntactic) records, the record name actually has uses (for inheritence, for example), but I don't see that it helps for what we are discussing here.

The issue lies fundamentally deeper, I think. In R7RS and (by default) in R6RS, record types are generative. In particular, their types do not exist at expand time. On the other hand, for type inference, Steme will probably need a static set of types.

So it may be that we have algebraic data types for Steme on the one hand side and Scheme record types and enumerations on the other hand, but that a bridge between both can only be provided by the FFI on a case-by-case basis.

**Comment by johnwcowan** [comment]
I don't think it's as bad as that.   A Steme constructor can still be made visible to Scheme as a Scheme record type, and in the reverse direction, Steme could be made to see record types as constructors, provided they are global.  We already disallow the exporting of anything but globally (to the library) defined identifiers, with good reason.

**Comment by mnieper** [comment]
If we are talking about R7RS, the bindings created by `define-record-type` have no special status as R7RS neither has inheritence nor inspection for record types. It is therefore not clear what is meant by "as a Scheme record type". But this makes things actually easier. Every Steme constructor can just become a Scheme procedure.

I still think that the reverse direction is more complicated. The bindings of even a top-level record type become only visible after the library is evaluated because they are variable bindings. This, however, will be too late for compiling Steme code against it. For R6RS nongenerative records, this is a bit easier when the record name is an expand-time object.

But I wouldn't worry too much about it at this stage. We can later always add syntax to Scheme that creates counterparts of Steme types.

**Comment by johnwcowan** [comment]
I'm not referring to the binding for the record type itself, which serves no purpose in R7RS-small.  I'm referring to the associated bindings.  For example, let's say we have written the following type definition:

```
(define-datatype Foo
  (bar (String a))
  (baz (String a) (String b)))
```

If this datatype is exported, it causes the following names to be importable into Scheme:  `Foo?`, `bar?`, `baz?`,  `make-bar`, `make-baz`, `bar-a`, `baz-a`, `baz-b`.  That allows Scheme full read-only access to any Steme object of type Foo.  

In addition, `bar` and `baz` can be made importable, but there is nothing to do with them until we make decisions about extension and/or introspection.  (I am for introspection facilities and against extension, FWIW.)  `Foo` is not importable because it is not reified and therefore not a Scheme type.

**Comment by mnieper** [comment]
Yes, as I said, this direction is much more straightforward. It is the other one (Scheme record types to Steme types) that will probably need some more work (e.g. through FFI annotations).

Just one note about `foo?` or `bar?` in your example: The status of these predicates on the Steme side is not yet clear to me. On the Scheme side, they are probably meant as procedures from all objects to the booleans, which is what would be expected. On the Steme side, we give a type to such a procedure. So they could only be functions from Foo to the booleans. But then they won't have a lot to do with the Scheme counterparts.

**Comment by aijony** [comment]
> On the Scheme side, they are probably meant as procedures from all objects to the booleans, which is what would be expected. On the Steme side, we give a type to such a procedure. So they could only be functions from Foo to the booleans. 

This should work because `bar?` will just take any object (possibly check if it is `Foo?` during runtime)¹ and check if it is `bar`, but it's type would be `(-> Foo Boolean)` so it would need to type check in Steme.

[1] This can for sure be optimized out of Steme if things type-check, and most likely won't be needed even for Scheme.

**Comment by mnieper** [comment]
I agree that this would be the right approach on the Steme side. But it shows that we won't get a very well-behaved correspondence between Steme sum types on the one hand side and "types" defined by a bunch of disjoint predicates on the Scheme side (where a predicate is a total function evaluating to a boolean).

**Comment by mnieper** [comment]
PS As we have been discussing records here, let me add one proposal, namely I would add two types of product types:

- Value products
- Computational products

Value products are finite tuples, completely evaluated, and fit best to the union approach we have discussed here, i.e. every such tuple type is part of an encompassing union.

Computation products, on the other hand, can model infinitary types like streams or generators. Creating such a struct means that the evaluation of the components will be delayed until they are accessed through some projector. In particular, computational structs of even infinite length (meaning with infinitely many fields) would be theoretically possible.

The natural domain of a `fold`-like procedure would be values of the first kind. The natural codomain of an `unfold`-like procedure would be computations of the second kind.

**Comment by aijony** [comment]
> I agree that this would be the right approach on the Steme side. But it shows that we won't get a very well-behaved correspondence between Steme sum types on the one hand side and "types" defined by a bunch of disjoint predicates on the Scheme side (where a predicate is a total function evaluating to a boolean).

Good point, I wouldn't be opposed to predicates (e.g. `bar?`) giving an error on the Scheme side of things (if that's what you mean).
